R/expect-setequal.R

```diff
@@ -78,9 +78,32 @@ expect_mapequal <- function(object, expected) {
     abort("`object` and `expected` must both be vectors")
   }
 
+  # Length-0 vectors are OK whether named or unnamed.
+  if (length(act$val) == 0 && length(exp$val) == 0) {
+    succeed()
```

Comment:
`succeed()` doesn't return so you also need `return(invisible(act$val))` here.

But are you sure this shouldn't be an error? It seems unlikely to be what you actually want.

Reply:
I'd expect this comparison to work (it's in the tests below):

```
x <- list()
x
#> list()

y <- list(a=1)[0]
y
#> named list()

expect_setequal(x, y)
```

And also:

```R
expect_setequal(list(), list())
```

When using a list as simple map or key-value store, I sometimes start with a `list()` and then populate it, and R automatically upgrades it to a named list. If that named list is then reduced to length 0, I'd expect this expectation to treat it as the same as an empty unnamed list.

Reply:
I think it make sense -- we'd expect `expect_setequal(character(), character())` to succeed; this is a similar analogue I think. (I also don't think the distinction between an unnamed vector and an empty vector with 0-length names is meaningful in this context)

Reply:
Yeah, I think `mapequal(list(), list())` should be `TRUE`, but it's not clear to me whether the same reasoning applied to `expect_mapequal()`. Because if you are comparing two empty lists, you really should be using something else. I think this would be at least worth a warning.

Reply:
Oh, I meant to write `expect_mapequal()`, not `expect_setequal()`. At any rate, I'm OK with a warning here.